### PR TITLE
FEATURE: Preload resources via link header

### PR DESCRIPTION
**Pull request by xfalcox**
Experiment moving from preload tags in the document head to preload information the the response headers.

While this is a minor improvement in most browsers (headers are parsed before the response body), this allows smart proxies like Cloudflare to "learn" from those headers and build HTTP 103 Early Hints for subsequent requests to the same URI, which will allow the user agent to download and parse our JS/CSS while we are waiting for the server to generate and stream the HTML response.

Before
waterfall(1)

After
waterfall

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1008,4 +1009,14 @@ def run_second_factor!(action_class, action_data = nil)
 
     result
   end
+
+  def link_preload
```
Comment (xfalcox) [inline app/controllers/application_controller.rb:1013]: Should we cache this for a couple of minutes?

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -21,8 +21,8 @@
 
     <%= build_plugin_html 'server:before-script-load' %>
 
-    <link rel="preload" href="<%= script_asset_path "start-discourse" %>" as="script">
-    <link rel="preload" href="<%= script_asset_path "browser-update" %>" as="script">
+    <% add_resource_preload_list(script_asset_path("start-discourse"), "script") %>
+    <% add_resource_preload_list(script_asset_path("browser-update"), "script") %>
```
Comment (tgxworld) [inline app/views/layouts/application.html.erb:25]: On browsers that do not yet support early hints, do we lose the preload with the link tag?

Comment (xfalcox) [reply app/views/layouts/application.html.erb:25]: No, we do not as they are interchangeable

Comment (tgxworld) [reply app/views/layouts/application.html.erb:25]: O nice, I didn't know about this.

File: app/helpers/application_helper.rb [hunk 3]
```diff
@@ -142,12 +142,16 @@ def preload_script(script)
...
   def preload_script_url(url)
+    add_resource_preload_list(url, 'script')
     <<~HTML.html_safe
-      <link rel="preload" href="#{url}" as="script">
       <script defer src="#{url}"></script>
     HTML
   end
 
+  def add_resource_preload_list(resource_url, type)
+    @links_to_preload << %Q(<#{resource_url}>; rel="preload"; as="#{type}") if !@links_to_preload.nil?
```
Comment (pmusaraj) [inline app/helpers/application_helper.rb:152]: Can we add some basic unit tests for this method? Mostly as documentation.

File: lib/stylesheet/manager.rb [hunk 4]
```diff
@@ -196,10 +196,11 @@ def stylesheet_preload_tag(target = :desktop, media = 'all')
     end.join("\n").html_safe
   end
 
-  def stylesheet_link_tag(target = :desktop, media = 'all')
+  def stylesheet_link_tag(target = :desktop, media = 'all', preload_callback = nil)
     stylesheets = stylesheet_details(target, media)
     stylesheets.map do |stylesheet|
       href = stylesheet[:new_href]
+      preload_callback.call(href, 'style') if preload_callback
```
Comment (pmusaraj) [inline lib/stylesheet/manager.rb:203]: This needs a test as well, current `stylesheet_link_tag` won't cover the preload callbacks.

File: spec/helpers/application_helper_spec.rb [hunk 5]
```diff
@@ -61,36 +60,77 @@ def preload_link(url)
...
         helper.request.env["HTTP_ACCEPT_ENCODING"] = 'gzip, br'
         link = helper.preload_script('discourse/tests/theme_qunit_ember_jquery')
-        expect(link).to eq(preload_link("https://s3cdn.com/assets/discourse/tests/theme_qunit_ember_jquery.js"))
+        expect(link).to eq(script_tag("https://s3cdn.com/assets/discourse/tests/theme_qunit_ember_jquery.js"))
       end
     end
   end
 
+  describe "add_resource_preload_list" do
+    it "adds resources to the preload list when it's availiable" do
```
Comment (pmusaraj) [inline spec/helpers/application_helper_spec.rb:94]:
```suggestion
    it "adds resources to the preload list when it's available" do
```

**Review by pmusaraj — APPROVED** [review]
Thanks!

**Pull request merged**